### PR TITLE
[SPARK-19525][CORE]Add RDD checkpoint compression support

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

This PR adds RDD checkpoint compression support and add a new config `spark.checkpoint.compress` to enable/disable it. Credit goes to @aramesh117

Closes #17024

## How was this patch tested?

The new unit test.

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @mridulm since you reviewed the initial PR.

**Comment by aramesh117** [comment]
@zsxwing Sorry for the delay! Thank you so much for your review and I saw a bit of your patch - it looks very nice. I have just one question - would it be a good idea to separate the codecs for compressing checkpoints and for network communication? I see that you reuse the same codec. If we wanted to isolate the effect of the RDD checkpointing, we would not be able to do that easily if these are coupled.

File: core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala [hunk 1]
```diff
@@ -169,7 +177,12 @@ private[spark] object ReliableCheckpointRDD extends Logging {
     val bufferSize = env.conf.getInt("spark.buffer.size", 65536)
 
     val fileOutputStream = if (blockSize < 0) {
-      fs.create(tempOutputPath, false, bufferSize)
+      val fileStream = fs.create(tempOutputPath, false, bufferSize)
+      if (env.conf.get(CHECKPOINT_COMPRESS)) {
+        CompressionCodec.createCodec(env.conf).compressedOutputStream(fileStream)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/rdd/ReliableCheckpointRDD.scala]: A question I had even with the earlier PR was - should we add the extension to either the directory or the file indicating compression type ?

**Review by mridulm — APPROVED** [review]
LGTM, thanks for pushing on this !
I have a query regarding naming though in case codec's are enabled.

**Comment by zsxwing** [comment]
> A question I had even with the earlier PR was - should we add the extension to either the directory or the file indicating compression type ?

Shuffle and cache files don't have an extension. I think it's better to be consistent in the whole code path.

> would it be a good idea to separate the codecs for compressing checkpoints and for network communication?

Save as above. Shuffle and cache files use the same codec.

**Comment by zsxwing** [comment]
In addition, I agree that having an extension and separating the codecs are good ideas. But they should be done in other PRs to not introduce multiple features in a large PR.

**Comment by mridulm** [comment]
Shuffle and cache files are not on hdfs :-) They do not potentially survive the application or be consumed OOB for recovery/inspection.

**Comment by mridulm** [comment]
Sounds good on doing it in separate PR - I am not too worried about shuffle/blockdata/etc btw - since they are private to application execution - checkpoint's tend to also be perused for other purposes (whether by design or not) since they are on hdfs; and with this PR it will become necessary to either guess the codec by iterating over supported codec's or pass it through other means for consumers.

**Comment by zsxwing** [comment]
Streaming checkpoints are on HDFS but don't have an extension :)

**Comment by mridulm** [comment]
@zsxwing They are compressed ? Interesting ... I never played with spark streaming unfortunately, so did not know !

**Comment by zsxwing** [comment]
yes. See https://github.com/apache/spark/blob/master/streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala#L138

**Comment by mridulm** [comment]
I thought the main reason @aramesh117 did this PR was for compression to be enabled for spark streaming usecase.
If compression is already enabled, then am I missing something here ?

**Comment by mridulm** [comment]
To add, for non streaming usecases, this will definitely help - but was this a recent change for streaming ? (probably after @aramesh117 make the PR ?)

**Comment by zsxwing** [comment]
> did this PR was for compression to be enabled for spark streaming usecase.

Streaming checkpoint includes two parts:
- DStream graph and metadata
- RDD checkpoints

Right now the first one is compressed. This PR is for RDD checkpoints.

**Comment by mridulm** [comment]
Ah interesting, thanks for clarifying ... weird that first was compressed and not second. But if there is an expectation some of the data is compressed already; perhaps we are being consistent now and there is no need to add extension (unless we uniformly do it everywhere).

LGTM, thanks for the change @zsxwing and @aramesh117 !
@zsxwing would it be possible for you to merge this ? I am facing some env issues right now. Thanks

**Comment by zsxwing** [comment]
Thanks, @mridulm  @aramesh117 Merging to master and 2.2.

**Comment by aramesh117** [comment]
@mridulm and @zsxwing thank you so much! This will help us out a lot! Much appreciated. :)

**Pull request closed without merging**